Fixes #8046: Add basic infoblox dhcp support

Klaas- opened the pull request:
This should add basic dhcp support for infoblox appliances over the rest api using the infoblox gem >= 0.4.1 (https://github.com/govdelivery/infoblox)

unclebilly commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,145 @@
+      # TODO: Refactor this into the base class
+      raise InvalidRecord, "#{record} is static - unable to delete" unless record.deleteable?
+      # "Deleting"" a record here means just disabling dhcp
+      connection = ::Infoblox::Connection.new(:username => Proxy::DhcpPlugin.settings.username, :password => Proxy::DhcpPlugin.settings.password, :host => Proxy::DhcpPlugin.settings.dhcp_server)
```
You might consider creating a `connection` method... this line is repeated several times in different methods.

Klaas- replied:
okay :)

lzap commented:
Hello,

shell we test against this product?

https://www.infoblox.com/downloads/software/ddi-trial

Can you specify supported products and versions in the example conf
file?

## 

Later,
 Lukas #lzap Zapletal

dmitri-d commented:
@Klaas-: could you add tests for this please?

dmitri-d commented:
[test]

domcleal commented:
The test failures are for style checks, which you can see at http://ci.theforeman.org/job/test_proxy_develop_pr_rubocop/276/checkstyleResult/ or by running `rubocop` locally.

Commit:
Fixes #8046: Add basic infoblox dhcp support

Klaas- commented:
Hi,
was busy on another project - I fixed the style checks, my local rubocop complains about more than jenkins - I think I fixed all style problems.
@lzap all infoblox products support their wapi layer -- so yeah I think you can try that :)
@witlessbird what kind of tests do I need to write? How would one test them if they don't have an infoblox application? Do I need to write a mock implementation of the restapi?

dmitri-d commented:
[test]

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+        # if not empty, first element is what we want to edit
+        host = host.first
+        post = false
+      end
```
I'd break this up into two methods one dealing with host creation, and another with modification.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      hostip.nextserver = options[:nextServer]
+      hostip.use_nextserver = true
+      hostip.bootfile = options[:filename]
+      hostip.use_bootfile = true
```
This bit can be extracted into a method too, which will help with the refactoring above.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      ## Test if Host Entry has correct IP
+      raise InvalidRecord, "#{record} IP mismatch" unless hostip.ipv4addr == record.ip
+      # Send object
+      post ? host.post : host.put
```
You won't need this conditional if you do the refactorings I suggested.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+       # Send object
+       host.put
+     end
+     subnet.delete(record)
```
You should call `super` here instead...

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+
+    def addRecord options={}
+      logger.debug "Add Record"
+      record = super(options)
```
No need to pass arguments here, `record = super` is sufficient.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      logger.debug "loadRecord"
+      connection = connect
+      # if record is a String it can be either ip or mac, true = mac --> lookup ip
+      if record.is_a?(String) && (IPAddr.new(record) rescue nil).nil?
```
You can use https://github.com/theforeman/smart-proxy/blob/develop/lib/proxy/validations.rb#L9 to check if a string is a valid mac address.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+        hostdhcp = ::Infoblox::HostIpv4addr.find(connection, "mac" => record).first
+        ipv4address = hostdhcp.ipv4addr
+      elsif record.is_a?(String)
+        ipv4address = record
```
probably worth validating this.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+        ipv4address = record
+      end
+      ipv4address = record[:ip] if record.is_a?(Proxy::DHCP::Record)
+      ipv4address = record.to_s if record.is_a?(IPAddr)
```
I'm quite certain `record` parameter in `find_record` is never a `Proxy::DHCP::Record` or an `IPAddr`, only a string.

Klaas- replied:
thats what server.rb does for find_record -- during my tests I only saw strings :) if you're certain about that we can maybe remove it from both? :)

dmitri-d replied:
I'm working on this in a dedicated PR.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      connection = connect
+      # if record is a String it can be either ip or mac, true = mac --> lookup ip
+      if record.is_a?(String) && (IPAddr.new(record) rescue nil).nil?
+        hostdhcp = ::Infoblox::HostIpv4addr.find(connection, "mac" => record).first
```
I don't know the api, but I assume that this will return a nil if dhcp record cannot be found? You'll need to handle that. Ditto for `::Infoblox::Host.find` call below

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      ipv4address = record.to_s if record.is_a?(IPAddr)
+      host = ::Infoblox::Host.find(connection, "ipv4addr" => ipv4address).first
+      return nil if host.nil? || host.name.empty?
+      hostdhcp = ::Infoblox::HostIpv4addr.find(connection, "ipv4addr" => ipv4address).first
```
Do you need to make this call when `find_record` is called with a mac address? You should be able to use the results of the call on line 85...

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+    end
+    
+    def find_record(record)
+      logger.debug "loadRecord"
```
you should call `super` here, and check if the record has already been loaded. Otherwise you'll get an error at the end when attempting to create a record that already exists.

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      # Load network from infoblox, iterate over ips to gather additional settings
+      logger.debug "LoadSubnetData"
+      super
+      network = IPAddr.new(subnet.to_s, Socket::AF_INET)
```
can't you use `subnet.network` here instead?

dmitri-d commented on modules/dhcp/providers/server/infoblox.rb:
```diff
@@ -0,0 +1,148 @@
+      network.each do |host|
+        # next if certain values are not set
+        next if host.names.empty? || host.mac_address.empty? || host.ip_address.empty?
+        hostdhcp = ::Infoblox::HostIpv4addr.find(connection, "ipv4addr" => host.ip_address).first
```
I did a quick look at infoblox api, but couldn't find any bulk operations, is that true? You may want to check if hostdhcp isn't empty -- if the number of records is large and leases short, there's a chance that the record may not be there by the time you get here.

dmitri-d commented:
@Klaas- 

for https://github.com/theforeman/smart-proxy/pull/295/files#diff-5daf377f40c93007066c8742a07aa442R20 I'd have a test that verifies that `::Infoblox::Host.find` is made with correct arguments, a test to make sure host gets updated, and a test to verify that the dhcp record is removed from the subnet.

for https://github.com/theforeman/smart-proxy/pull/295/files#diff-5daf377f40c93007066c8742a07aa442R40 (assuming you do the refactorings): test to verify search call uses correct args, test to verify host record creation parameters, test to verify common method that updates host record, test to verify record creation uses post, and test to verify record modification uses put.

for https://github.com/theforeman/smart-proxy/pull/295/files#diff-5daf377f40c93007066c8742a07aa442R80: test that ::Infoblox::HostIpv4addr.find is called with correct args when looking up by mac/ip, and another test to verify that `Proxy::DHCP::Record` is being created with correct values.

for https://github.com/theforeman/smart-proxy/pull/295/files#diff-5daf377f40c93007066c8742a07aa442R108: test that ::Infoblox::Ipv4address.find on line 115 is made with correct args, then another to test that `Proxy::DHCP::Reservation`s are being populated correctly (for the sake of simplicity I'd return a single host in the network.

You'll need to use mocks/stubs here quite a bit, we use mocha (http://gofreerange.com/mocha/docs/). Feel free to ask me any questions you have re: testing, mocking, or comments I made.

ohadlevy commented:
@Klaas- ping?

Klaas- commented:
I'm busy doing something else atm, expect some more development on this 
in 1-2 months, in the mean time all help and further improvement ideas 
are welcome :)

dmitri-d commented:
I doubt now that dhcp provider modularization is largely completed this PR will be accepted into the core. I would suggest releasing support for infoblox as a module. PR that converts existing dhcp providers into modules is here: https://github.com/theforeman/smart-proxy/pull/312. You can use that as a starting point, and I'm available to answer questions should you have any.

MagnusJernberg commented:
@Klaas- @witlessbird 
Hello, i'm looking into different ipam tools and are also interested in infoblox. we need foreman to work with a 3rd party ipam tool. 

Does the previous comments mean that this fix will not be merged and that support for infoblox needs to be added as a part of #312 ?

dmitri-d commented:
@MagnusJernberg: not part of https://github.com/theforeman/smart-proxy/pull/312, but rather a separate PR that implements infoblox as a provider.

dmitri-d commented:
Closing this PR, as infoblox support should be implemented as a dhcp module provider. @Klaas-: let me know if you need some pointers for how to create a provider...
